libcaf_core/caf/flow/op/buffer.test.cpp

```diff
@@ -0,0 +1,499 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#include "caf/flow/op/buffer.hpp"
+
+#include "caf/test/fixture/deterministic.hpp"
+#include "caf/test/fixture/flow.hpp"
+#include "caf/test/scenario.hpp"
+#include "caf/test/test.hpp"
+
+#include "caf/all.hpp"
+
+using namespace caf;
+using namespace std::literals;
+
+namespace {
+
+constexpr auto fwd_data = caf::flow::op::buffer_input_t{};
+
+constexpr auto fwd_ctrl = caf::flow::op::buffer_emit_t{};
+
+struct skip_trait {
+  static constexpr bool skip_empty = true;
+  using input_type = int;
+  using output_type = cow_vector<int>;
+  using select_token_type = int64_t;
+
+  output_type operator()(const std::vector<input_type>& xs) {
+    return output_type{xs};
+  }
+};
+
+struct noskip_trait {
+  static constexpr bool skip_empty = false;
+  using input_type = int;
+  using output_type = cow_vector<int>;
+  using select_token_type = int64_t;
+
+  output_type operator()(const std::vector<input_type>& xs) {
+    return output_type{xs};
+  }
+};
+
+struct fixture : test::fixture::deterministic, test::fixture::flow {
+  template <class Impl>
+  void add_subs(intrusive_ptr<Impl> uut) {
+    auto data_sub
+      = make_counted<test::fixture::flow::passive_subscription_impl>(
+        coordinator());
+    uut->fwd_on_subscribe(fwd_data,
+                          caf::flow::subscription{std::move(data_sub)});
+    auto ctrl_sub
+      = make_counted<test::fixture::flow::passive_subscription_impl>(
+        coordinator());
+    uut->fwd_on_subscribe(fwd_ctrl,
+                          caf::flow::subscription{std::move(ctrl_sub)});
+  }
+
+  template <class T>
+  auto trivial_obs() {
+    return test::fixture::flow::make_trivial_observable<T>(coordinator());
+  }
```

Comment:
I wonder if we really need this. I think the tests would work just as well with `never` and wouldn't require us to hand-roll a new observable implementation for testing.